### PR TITLE
[SPARK-49249][SPARK-49122] Artifact isolation in Spark Classic

### PR DESCRIPTION
**Pull request by xupefei**
### What changes were proposed in this pull request?

This PR makes the isolation feature introduced by `SparkSession.addArtifact` API (added in https://github.com/apache/spark/pull/47631) work with Spark SQL.

Note that this PR does not enable isolation for the following two use cases:

- PySpark
	- Future work is needed to add API to support adding isolated Python UDTFs.
- When Hive is used as the metastore
	- Hive UDF is a huge blocker due to artifacts can be used outside a `SparkSession`, which resources escaped from our session scope.

### Why are the changes needed?

Because it didn't work before :)

### Does this PR introduce _any_ user-facing change?

Yes, the user can add a new artifact in the REPL and use it in the current REPL session.

### How was this patch tested?

Added a new test.

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/UDFRegistration.scala
Comment (xupefei) [inline sql/core/src/main/scala/org/apache/spark/sql/UDFRegistration.scala:150]: I have to make this method public so I can call it from REPL.

Comment (hvanhovell) [reply sql/core/src/main/scala/org/apache/spark/sql/UDFRegistration.scala:150]: I am not against this. I am trying to understand the user facing consequences though. I'd probably prefer that we add support for Scala UDFs as well. That can be done in a follow-up though.

File: sql/core/src/main/scala/org/apache/spark/sql/UDFRegistration.scala
Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/UDFRegistration.scala:125]: We could probably be a bit more direct here, by using the classloader produced by the artifactManager directly.

File: repl/src/test/scala/org/apache/spark/repl/ReplSuite.scala
Comment (hvanhovell) [inline repl/src/test/scala/org/apache/spark/repl/ReplSuite.scala:400]: Can you use a UDF defined in the REPL?  If so how does this work with a JobArtifactSet? Do we layer the globally defined classpath over the session specific classpath? (I'd be nice to document this somewhere).

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala
Comment (hvanhovell) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala:118]: An FYI to other reviewers: look at this file with hidden whitespace.

Comment (vicennial) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala:118]: Hmm, with this in the execution code path, we may not need SessionHolder#withSession in a few places and can be cleaned up.

**Review by vicennial — APPROVED** [review]

**Review by vicennial — COMMENTED** [review]
Oops, I meant to leave a comment (ignore the approval, I haven't gone through the whole PR)

File: repl/src/test/scala/org/apache/spark/repl/ReplSuite.scala
Comment (xupefei) [reply repl/src/test/scala/org/apache/spark/repl/ReplSuite.scala:400]: I added one more test, which defines a UDF that initialises an external class added as an artifact.

> how does this work with a JobArtifactSet?

Can you elaborate? Afaik `JobArtifactSet` is not involved here since it's the artifact path that is applied when an active SparkSession is applied.

Classpath - It's the other way around: the session classpath is laid over the global one.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala
Comment (xupefei) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala]: @vicennial Is there a end-to-end test for this? I did some modifications and want to know if it won't break anything.

File: sql/core/src/main/scala/org/apache/spark/sql/UDFRegistration.scala
Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/UDFRegistration.scala:123]: One follow-up here would be to cache the ArtifactManager classloader. I think we create that thing over and over.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala
Comment (vicennial) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala]: @xupefei The [ReplE2ESuite](https://github.com/apache/spark/blob/dee53bb31edd877799cfe06473ec3a47a42e6774/connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/application/ReplE2ESuite.scala#L34) has some tests for the overall client->artifact->execution with artifact flow. 
Python client package my have some E2E tests as well but I am not familiar of the current status.

File: sql/core/src/main/scala/org/apache/spark/sql/UDFRegistration.scala
Comment (xupefei) [reply sql/core/src/main/scala/org/apache/spark/sql/UDFRegistration.scala:123]: Changed it to a `lazy val`. It is safe because during initialisation it only reads read-only SparkEnv configs

Comment (xupefei) [reply sql/core/src/main/scala/org/apache/spark/sql/UDFRegistration.scala:123]: Agree. We can invalid the cache when a new JAR is added.

File: sql/core/src/main/scala/org/apache/spark/sql/UDFRegistration.scala
Comment (hvanhovell) [reply sql/core/src/main/scala/org/apache/spark/sql/UDFRegistration.scala:151]: Can you file a follow-up?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala
Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala:284]: I think it should be safe to use the SparkSession's jobArtifactState. They should be the same. cc @vicennial.

**Review by hvanhovell — APPROVED** [review]
LGTM - Pending @vicennial's sign-off.

File: sql/core/src/main/scala/org/apache/spark/sql/UDFRegistration.scala
Comment (xupefei) [reply sql/core/src/main/scala/org/apache/spark/sql/UDFRegistration.scala:151]: Will do.

File: sql/core/src/main/scala/org/apache/spark/sql/artifact/ArtifactManager.scala
Comment (xupefei) [inline sql/core/src/main/scala/org/apache/spark/sql/artifact/ArtifactManager.scala:80]: We have to fully turn off resource isolation not to break these magical classpath manipulations:

1. REPL inline artifacts folder `spark.repl.class.outputDir` and `spark.repl.class.uri`.
2. Hive-related JARS loaded from `hive.aux.jars.path`: https://github.com/apache/spark/blob/80d6651cf6a1835d0de3e12e08253d2a9816d499/sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkSQLCLIDriver.scala#L168-L172

File: sql/core/src/main/scala/org/apache/spark/sql/artifact/ArtifactManager.scala
Comment (vicennial) [inline sql/core/src/main/scala/org/apache/spark/sql/artifact/ArtifactManager.scala:92]: Can we add a short blurb on when this boolean should be set to true?

**Review by vicennial — COMMENTED** [review]
Overall LGTM for the logic % few nits mentioned above

File: sql/core/src/main/scala/org/apache/spark/sql/artifact/ArtifactManager.scala
Comment (xupefei) [reply sql/core/src/main/scala/org/apache/spark/sql/artifact/ArtifactManager.scala:87]: No, they can be different in other sessions.

File: sql/core/src/main/scala/org/apache/spark/sql/artifact/ArtifactManager.scala
Comment (xupefei) [reply sql/core/src/main/scala/org/apache/spark/sql/artifact/ArtifactManager.scala:92]: Done!

File: sql/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectServer.scala
Comment (hvanhovell) [inline sql/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectServer.scala:34]: This is the default right?

File: sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala
Comment (hvanhovell) [reply sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala:843]: Let's file a follow-up to make this work in Hive.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala
Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala:123]: Can we remove this in the connect server since we are doing it here?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala
Comment (hvanhovell) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala:308]: Ping?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala
Comment (hvanhovell) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala:754]: Why is this change needed?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala
Comment (hvanhovell) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala:2144]: Can you try to construct the path instead? This feels a bit too much like a work around?

File: sql/core/src/main/scala/org/apache/spark/sql/artifact/ArtifactManager.scala
Comment (xupefei) [inline sql/core/src/main/scala/org/apache/spark/sql/artifact/ArtifactManager.scala:321]: This classloader is used only in Hive to hold all `ADD JAR`s.

This classloader will be set to default once there's one ADD JAR: https://github.com/apache/spark/blob/6a36c43604a638f603a3a40c22ee1e6bd3ae8d7e/sql/core/src/main/scala/org/apache/spark/sql/internal/SessionState.scala#L188

File: sql/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectServer.scala
Comment (xupefei) [reply sql/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectServer.scala:34]: Yes. I removed this line.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala
Comment (xupefei) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala:754]: My bad. This is a leftover from me poking around trying to optimise the performance.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala
Comment (xupefei) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala:123]: Yes.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:3949]: Why do we need this flag? It seems that we should be as lazy as possible here.

File: sql/connect/server/src/main/scala/org/apache/spark/sql/connect/SimpleSparkConnectService.scala
Comment (hvanhovell) [reply sql/connect/server/src/main/scala/org/apache/spark/sql/connect/SimpleSparkConnectService.scala:44]: So why are we setting it?

File: sql/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectServer.scala
Comment (hvanhovell) [reply sql/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectServer.scala:34]: What did you remove?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala
Comment (hvanhovell) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala:2144]: Is add file isolated? If not, then why do we need this?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala
Comment (xupefei) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:3949]: This flag is solely used by Connect, to not break any existing test case and allow it to access artifacts in the "default" workspace before the first `adArtifact` call.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala
Comment (vicennial) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:3947]: What is the behaviour when some sessions have this disabled, and some have this enabled? Will all the sessions which disabled this flag share resources amongst each other?

File: sql/core/src/main/scala/org/apache/spark/sql/artifact/ArtifactManager.scala
Comment (vicennial) [inline sql/core/src/main/scala/org/apache/spark/sql/artifact/ArtifactManager.scala:80]: Note: Let's make sure to mention somewhere changing the config in the middle of session usage is a no-op.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala
Comment (xupefei) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala:2144]: Now it's hacky: The JAR will be added to SparkContect with a session UUID prefix but will be accessible by all sessions (due to we wrapped it's classloader).

On a second thought, I could make it cleaner by passing `default` as UUID into the `sparkContext.addJar` method to explicitly disable isolation.

Comment (xupefei) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala:2144]: This is a leftover from another PR that enables Hive isolation. Removed!

File: repl/src/main/scala/org/apache/spark/repl/Main.scala
Comment (xupefei) [reply repl/src/main/scala/org/apache/spark/repl/Main.scala:99]: Done!

File: sql/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectServer.scala
Comment (xupefei) [reply sql/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectServer.scala:34]: I removed setting this flag before but some connect tests are failing afterwards, seems not to get the default value. Now I added it back to make sure the isolation is indeed enabled.

File: sql/core/src/main/scala/org/apache/spark/sql/UDFRegistration.scala
Comment (xupefei) [reply sql/core/src/main/scala/org/apache/spark/sql/UDFRegistration.scala:123]: Created https://issues.apache.org/jira/browse/SPARK-50243.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala
Comment (xupefei) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala:308]: I'll do the cleanup as a follow-up.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala
Comment (xupefei) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:3947]: Yes, they'll be sharing artifacts under the `default` namespace.

File: sql/core/src/main/scala/org/apache/spark/sql/artifact/ArtifactManager.scala
Comment (xupefei) [reply sql/core/src/main/scala/org/apache/spark/sql/artifact/ArtifactManager.scala:80]: Done! I rephased the doc to
```
.doc("When enabled for a Spark Session, artifacts (such as JARs, files, archives) added to " +
  "this session are isolated from other sessions within the same Spark instance. When " +
  "disabled for a session, artifacts added to this session are visible to other sessions " +
  "that have this config disabled. This config can only be set during the creation of a " +
  "Spark Session and will have no effect when changed in the middle of session usage.")
```

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala
Comment (xupefei) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala]: I reverted this change to left it to future (when enabling isolation for Hive).

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala
Comment (hvanhovell) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:3952]: Which test is failing? It is a bit weird to me that we need this for connect.

Comment (xupefei) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:3952]: This: https://github.com/xupefei/spark/actions/runs/11628160404/job/32382826661
It checks for the error message from the StubClassLoader.

**Review by vicennial — APPROVED** [review]
LGTM!

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala
Comment (xupefei) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala]: I found that my previous comment was wrong. `ADD JAR` is isolated when used outside of Hive.
To not bring a behaviour change, I'll make a small change to disable isolation for this command even outside Hive.

Comment (xupefei) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala]: And we can discuss whether we should re-enable isolation as a follow-up.

File: sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala
Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala:847]: @xupefei Everyone has Hive on their classpath (because of HMS), so this effectively disables isolation for all of classic. Can we only disable this for JDBC?

File: sql/core/src/main/scala/org/apache/spark/sql/artifact/ArtifactManager.scala
Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/artifact/ArtifactManager.scala:76]: A downside of using a lazy val here, is that the moment when determine this might not be obvious to the user.

File: sql/core/src/main/scala/org/apache/spark/sql/artifact/ArtifactManager.scala
Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/artifact/ArtifactManager.scala:95]: I still don't understand why we need to set a classloader when there are no artifacts added?

File: sql/core/src/main/scala/org/apache/spark/sql/artifact/ArtifactManager.scala
Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/artifact/ArtifactManager.scala:89]: The name of the value is a bit misleading. It signals when the classpath is modified, not when any artifact is added.

File: sql/core/src/main/scala/org/apache/spark/sql/artifact/ArtifactManager.scala
Comment (hvanhovell) [reply sql/core/src/main/scala/org/apache/spark/sql/artifact/ArtifactManager.scala:292]: mmm'kay... not loving it....

**Review by hvanhovell — APPROVED** [review]
LGTM

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master.

**Comment by HyukjinKwon** [comment]
Hm, I think actually it brow SparkR tests in Windows https://github.com/apache/spark/actions/workflows/build_sparkr_window.yml

Spark itself supports Windows so we should probably test it, or at least make the CI passes ....

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:3953]: Unfortunately, this looks like a violation of Apache Spark configuration rule. I made a follow-up.
- https://github.com/apache/spark/pull/49101

**Pull request closed without merging**
